src/awkward/_nplikes/__init__.py

```diff
@@ -27,6 +28,28 @@ def to_nplike(
     if from_nplike is nplike:
         return array
 
+    # We can always convert virtual arrays to typetracers
+    # but can only convert virtual arrays to other backends with known data only if they are intentionally materialized
+    # Only numpy and cupy nplikes are allowed for virtual arrays
+    if isinstance(array, awkward._nplikes.virtual.VirtualArray):
+        if not array.is_materialized and not isinstance(
+            nplike, awkward._nplikes.typetracer.TypeTracer
+        ):
+            raise TypeError(
+                "Cannot convert a VirtualArray to a different nplike with known data without materializing it first. Use ak.materialize on the array to do so."
+            )
+        else:
+            if isinstance(
+                nplike, (awkward._nplikes.numpy.Numpy, awkward._nplikes.cupy.Cupy)
+            ):
```

Comment:
I think here it might be nicer to have the `nplike` encode the "I support virtual arrays" part, e.g.

```python
class Numpy:
    supports_virtual_arrays = True
```

```suggestion
            if nplike.supports_virtual_arrays:
```

Reply:
Added in https://github.com/scikit-hep/awkward/pull/3364/commits/9fa5962c1cf284538edf282b6b200871ecc8fc56

Reply:
I like that! 👍